framework/include/utils/IndexableProperty.h

```diff
@@ -0,0 +1,154 @@
+ * instead of a `Real` material property to allow the user to supply any material
+ * property of a type from the list below along with a component index parameter to
+ * select a scalar component from the property value.
+ */
+template <typename T, bool is_ad>
+class IndexableProperty
+{
+public:
+  static InputParameters validParams();
+
+  IndexableProperty(T * host,
+                    const std::string & property_param = "property",
+                    const std::string & component_param = "component");
+
+  /// getthe selected component value for the given quadrature point
+  GenericReal<is_ad> operator[](int qp) const;
+
+  /// returns true if a supported material property has been found and can be evaluated
+  operator bool() const;
+
+protected:
+  template <typename P>
+  void getPropertyHelper(const GenericMaterialProperty<P, is_ad> *& pointer,
+                         unsigned int components);
+
+  ///@{ only one of those pointers will be non-null and pointing to the selected property
+  const GenericMaterialProperty<Real, is_ad> * _property_real;
+  const GenericMaterialProperty<std::vector<Real>, is_ad> * _property_std_vector;
+  const GenericMaterialProperty<RealVectorValue, is_ad> * _property_real_vector_value;
+  const GenericMaterialProperty<RankTwoTensor, is_ad> * _property_rank_two_tensor;
+  const GenericMaterialProperty<RankThreeTensor, is_ad> * _property_rank_three_tensor;
+  const GenericMaterialProperty<RankFourTensor, is_ad> * _property_rank_four_tensor;
+  ///@}
+
+  /// pointer to the host object
+  T * _host;
+
+  /// name of the input parameter containing the material property name
+  const std::string & _property_param;
+  /// name of the coupled material property (for error reporting)
+  const std::string & _property_name;
+  /// name of the input parameter containing the component index
+  const std::string & _component_param;
+
+  /// Index of the selected scalar component of the material property
+  const std::vector<unsigned int> _component;
+};
+
+template <typename T, bool is_ad>
+InputParameters
+IndexableProperty<T, is_ad>::validParams()
+{
+  auto params = T::validParams();
+  params.template addRequiredParam<MaterialPropertyName>("property",
+                                                         "The name of the material property");
+  params.template addParam<std::vector<unsigned int>>(
+      "component",
+      "Index vector of the scalar component to extract from "
+      "the material property (empty for scalar properties)");
+  return params;
+}
+
+template <typename T, bool is_ad>
+IndexableProperty<T, is_ad>::IndexableProperty(T * host,
+                                               const std::string & property_param,
+                                               const std::string & component_param)
+  : _host(host),
+    _property_param(property_param),
+    _property_name(_host->template getParam<MaterialPropertyName>(_property_param)),
```

Comment:
if you passed params instead of host couldn't you remove host?

Reply:
I had replied to that already

> Needed also for getParam and **getMateriaProperty** below

Reply:
true true. I was working my way to my last comment ^^
getMaterialProp comes along with the MatPropInterface